File: Backend/SanviiTechmet/src/main/java/com/nt/entity/LeaveRequest.java
package com.nt.entity;
import jakarta.persistence.Entity;
import jakarta.persistence.Column;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import lombok.Builder;
import java.time.LocalDateTime;
import java.util.Date;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Temporal;
import jakarta.persistence.TemporalType;
import jakarta.persistence.PreUpdate;
import org.springframework.data.annotation.Version;

@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class LeaveRequest {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long leaveRequestId;

    @Column(nullable = false)
    private Long userId; // Foreign key (Assuming Employee exists in another table)

    @Column(nullable = false)
    private Long reportingManagerId; //its exist in employee table

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date leaveStartDate;

    @Column(nullable = false)
    @Temporal(TemporalType.DATE)
    private Date leaveEndDate;

    @Column(nullable = false)
    private String leaveType;

    private String description;

    @Column(nullable = false)
    private String leaveStatus;

    private String createdBy;

    private LocalDateTime createdDate;

    private String updatedBy;

    private LocalDateTime updatedDate;


    @PrePersist
    protected void onCreate() {
        this.createdDate = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedDate = LocalDateTime.now();
    }
}

File: Nodejs backend/models/LeaveRequest.js
const mongoose = require('mongoose');

const leaveRequestSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  reportingManagerId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  leaveStartDate: { type: Date },
  leaveEndDate: { type: Date },
  leaveType: { type: String },
  description: { type: String },
  leaveStatus: { type: String, default: 'PENDING' },
  createdBy: { type: String },
  createdDate: { type: Date, default: Date.now },
  updatedBy: { type: String },
  updatedDate: { type: Date }
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

leaveRequestSchema.pre('findOneAndUpdate', function () {
  this.set({ updatedDate: new Date() });
});

module.exports = mongoose.model('LeaveRequest', leaveRequestSchema);
